src/lib/protocols/http.c

```diff
@@ -401,53 +405,127 @@ static void setHttpUserAgent(struct ndpi_detection_module_struct *ndpi_struct,
+    ndpi_unset_risk(ndpi_struct, flow, NDPI_KNOWN_PROTOCOL_ON_NON_STANDARD_PORT);
+  }
 
-    if(ndpi_struct->packet.iph
-       && (sscanf(flow->host_server_name, "%d.%d.%d.%d", &a, &b, &c, &d) == 4)) {
-      /* IPv4 */
+  /* Matching on Content-Type.
+      OCSP:  application/ocsp-request, application/ocsp-response
+  */
+  /* We overwrite any previous sub-classification (example: via hostname) */
+  if(packet->content_line.len > 17 &&
+     strncmp((const char *)packet->content_line.ptr, "application/ocsp-", 17) == 0) {
+    NDPI_LOG_DBG2(ndpi_struct, "Found OCSP\n");
+    ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_OCSP, master_protocol, NDPI_CONFIDENCE_DPI);
+  }
 
-      if(ndpi_struct->packet.iph->daddr != inet_addr(flow->host_server_name)) {
-	char buf[64], msg[128];
-	
-	snprintf(msg, sizeof(msg), "Expected %s, found %s",
-		 ndpi_intoav4(ntohl(ndpi_struct->packet.iph->daddr), buf, sizeof(buf)), flow->host_server_name);
-	ndpi_set_risk(ndpi_struct, flow, NDPI_HTTP_SUSPICIOUS_HEADER, msg);
-      }
-    }
-    
-    master_protocol = NDPI_PROTOCOL_HTTP;
-    if(flow->detected_protocol_stack[1] != NDPI_PROTOCOL_UNKNOWN)
-      master_protocol = flow->detected_protocol_stack[1];
-    else if(flow->detected_protocol_stack[0] == NDPI_PROTOCOL_HTTP_CONNECT ||
-            flow->detected_protocol_stack[0] == NDPI_PROTOCOL_HTTP_PROXY)
-      master_protocol = flow->detected_protocol_stack[0];
-
-    if(ndpi_match_hostname_protocol(ndpi_struct, flow,
-				    master_protocol,
-				    flow->host_server_name,
-				    strlen(flow->host_server_name)) == 0) {
-      if(flow->http.url &&
-         ((strstr(flow->http.url, ":8080/downloading?n=0.") != NULL)
-          || (strstr(flow->http.url, ":8080/upload?n=0.") != NULL))) {
-	/* This looks like Ookla speedtest */
-	ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_OOKLA, master_protocol, NDPI_CONFIDENCE_DPI);
-	ookla_add_to_cache(ndpi_struct, flow);
+  if(flow->http.method == NDPI_HTTP_METHOD_RPC_IN_DATA ||
+     flow->http.method == NDPI_HTTP_METHOD_RPC_OUT_DATA) {
+    ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_RPC, master_protocol, NDPI_CONFIDENCE_DPI);
+  }
+
+  if(flow->detected_protocol_stack[1] == NDPI_PROTOCOL_UNKNOWN &&
+     hostname_just_set && flow->host_server_name[0] != '\0') {
+    ndpi_match_hostname_protocol(ndpi_struct, flow,
+				 master_protocol,
+				 flow->host_server_name,
+				 strlen(flow->host_server_name));
+  }
+ 
+  if(flow->detected_protocol_stack[1] == NDPI_PROTOCOL_UNKNOWN &&
+     packet->http_origin.len > 0) {
+    ndpi_protocol_match_result ret_match;
+    char *ptr, *origin, *origin_hostname;
+
+    /* Origin syntax:
+        Origin: null
+        Origin: <scheme>://<hostname>
+        Origin: <scheme>://<hostname>:<port>
+    Try extracting hostname */
+
+    origin = ndpi_malloc(packet->http_origin.len + 1);
```

Comment:
Is it really required to use dynamic memory here?

Reply:
No, it isn't...

Reply:
Not sure if it is worth to replace it with a stack allocation. If I remember correctly allocating a variable sized buffer on the stack is a GNu extension, so MSVC may complain here.

Reply:
In the new version I simply used the original buffer, without any temporary/new buffer